### PR TITLE
Compare old results with new frontend api

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Running old frontend API and new frontend API side by side and comparing results.

Details:
* comparison only on when flag enabled
* always returning old results and warning if new results are different
* global cache is updated only when the flag is on
* copy pasted most of the old code with minor tweak so it's easy to delete old code after the migration
* sticking to the existing naming (separate PR will remove proxy and replace it with frontend API naming)
* next PR will create a composition root for the proxy service

### CODE REVIEW

File: src/lib/services/index.ts [hunk 1]
```diff
@@ -293,12 +296,27 @@ export const createServices = (
         ? createClientFeatureToggleService(db, config)
         : createFakeClientFeatureToggleService(config);
 
-    const proxyService = new ProxyService(config, stores, {
-        featureToggleServiceV2,
-        clientMetricsServiceV2,
-        settingService,
+    const clientFeatureToggleReadModel = db
```
Comment (kwasniew) [inline src/lib/services/index.ts:299]: this will be composition root in next pr

File: src/lib/services/proxy-service.ts [hunk 2]
```diff
@@ -85,6 +97,33 @@ export class ProxyService {
             }));
     }
 
+    async getNewProxyFeatures(
+        token: IApiUser,
+        context: Context,
+    ): Promise<ProxyFeatureSchema[]> {
+        const client = await this.newClientForProxyToken(token);
```
Comment (kwasniew) [inline src/lib/services/proxy-service.ts:104]: copy paste of old code with one line modification

File: src/lib/services/proxy-service.ts [hunk 3]
```diff
@@ -143,6 +196,33 @@ export class ProxyService {
         return client;
     }
 
+    private async createNewClientForProxyToken(
+        token: IApiUser,
+    ): Promise<Unleash> {
+        const repository = new FrontendApiRepository(
```
Comment (kwasniew) [inline src/lib/services/proxy-service.ts:202]: copy paste of old code with this line replaced to use new repository

File: src/lib/types/experimental.ts [hunk 4]
```diff
@@ -262,6 +263,10 @@ const flags: IFlags = {
         process.env.UNLEASH_EXPERIMENTAL_RESPONSE_TIME_METRICS_FIX,
         false,
     ),
+    globalFrontendApiCache: parseEnvVarBoolean(
```
Comment (kwasniew) [inline src/lib/types/experimental.ts:266]: new flag for safe rollout on the new frontend api

File: src/lib/services/proxy-service.ts [hunk 5]
```diff
@@ -117,6 +156,20 @@ export class ProxyService {
...
 
+    private async newClientForProxyToken(token: IApiUser): Promise<Unleash> {
+        ProxyService.assertExpectedTokenType(token);
+
+        let newClient = this.newClients.get(token.secret);
+        if (!newClient) {
+            newClient = this.createNewClientForProxyToken(token);
+            this.newClients.set(token.secret, newClient);
+            // TODO: do we need this twice?
+            // this.config.eventBus.emit(PROXY_REPOSITORY_CREATED);
```
Comment (kwasniew) [inline src/lib/services/proxy-service.ts:167]: we will uncomment this when the old client is gone

File: src/lib/services/proxy-service.ts [hunk 6]
```diff
@@ -45,21 +47,31 @@ export class ProxyService {
...
+
     /**
-     * This is intentionally a Promise becasue we want to be able to await
+     * This is intentionally a Promise because we want to be able to await
      * until the client (which might be being created by a different request) is ready
      * Check this test that fails if we don't use a Promise: src/test/e2e/api/proxy/proxy.concurrency.e2e.test.ts
      */
     private readonly clients: Map<ApiUser['secret'], Promise<Unleash>> =
         new Map();
+    private readonly newClients: Map<ApiUser['secret'], Promise<Unleash>> =
```
Comment (kwasniew) [inline src/lib/services/proxy-service.ts:59]: we store new and old clients side by side. when flag is disabled new clients will be empty.

File: src/lib/services/proxy-service.ts [hunk 7]
```diff
@@ -17,18 +17,20 @@ import { validateOrigins } from '../util';
 import { BadDataError, InvalidTokenError } from '../error';
 import { PROXY_REPOSITORY_CREATED } from '../metric-events';
 import { ProxyRepository } from '../proxy';
+import { FrontendApiRepository } from '../proxy/frontend-api-repository';
+import { GlobalFrontendApiCache } from '../proxy/global-frontend-api-cache';
 
-type Config = Pick<
+export type Config = Pick<
```
Comment (kwasniew) [inline src/lib/services/proxy-service.ts:23]: tests need those types

**Review by sjaanus — APPROVED** [review]
LG

**Pull request merged**